File: Day #1585/solution.py
# Day 1585: Sieve of Eratosthenes + incremental infinite prime generator.
# Sieve marks composites up to N; generator yields primes indefinitely via a sieve map.
# Time: O(N log log N) sieve; Space: O(N).
from itertools import islice


def sieve(n):
    comp = [False] * n
    primes = []
    for i in range(2, n):
        if not comp[i]:
            primes.append(i)
            for j in range(i * i, n, i):
                comp[j] = True
    return primes


def prime_generator():
    """Yield primes indefinitely (incremental sieve, no upper bound)."""
    composites = {}
    n = 2
    while True:
        if n not in composites:
            yield n
            composites[n * n] = [n]
        else:
            for p in composites.pop(n):
                composites.setdefault(n + p, []).append(p)
        n += 1


if __name__ == "__main__":
    print("Primes < 30:", sieve(30))
    print("First 10 primes:", list(islice(prime_generator(), 10)))

File: Day #1585/solution.go
// Day 1585: Sieve of Eratosthenes + incremental infinite prime generator.
// Sieve marks composites up to N; generator streams primes via a goroutine channel.
// Time: O(N log log N) sieve; Space: O(N).
package main

import "fmt"

func sieve(n int) []int {
	comp := make([]bool, n)
	var primes []int
	for i := 2; i < n; i++ {
		if !comp[i] {
			primes = append(primes, i)
			for j := i * i; j < n; j += i {
				comp[j] = true
			}
		}
	}
	return primes
}

// firstPrimes uses an incremental sieve (unbounded in principle).
func firstPrimes(count int) []int {
	var primes []int
	composites := map[int][]int{}
	for n := 2; len(primes) < count; n++ {
		if hit, ok := composites[n]; !ok {
			primes = append(primes, n)
			composites[n*n] = append(composites[n*n], n)
		} else {
			for _, p := range hit {
				composites[n+p] = append(composites[n+p], p)
			}
			delete(composites, n)
		}
	}
	return primes
}

func main() {
	fmt.Println("Primes < 30:", sieve(30))
	fmt.Println("First 10 primes:", firstPrimes(10))
}
